Switch from using docvalue_fields to extracting values from _source

astefan opened the pull request:
Doing this means parsing the `_source` and handling any fields that need to be parsed and have its outcome potentially changed by the parsers (for example, the numbers when `coerce=true`), just like Elasticsearch is doing it when it's indexing a document.

Fields that will not be taken from `_source` (but from `docvalue_fields`) are: `keyword`, `date`, `scaled_float`, `geo_point`, `geo_shape`.

This change also introduces a limitation: `alias`es type of fields that are NOT part of a tree of sub-fields will not be able to be retrieved anymore. `field_caps` API doesn't shed any light into a field being an
`alias` or not and at `_source` parsing time there is no way to know if aroot field is an alias or not. Fields of the type `a.b.c.alias` _can be_ extracted from `docvalue_fields`, only if the field they point to can be
extracted from `docvalue_fields`. Also, not all fields in a hierarchy of fields can be evaluated to being an alias.

Fixes https://github.com/elastic/elasticsearch/issues/41852.

astefan committed:
Switch from using docvalue_fields to extracting values from _source
where applicable. Doing this means parsing the _source and handling the
numbers parsing just like Elasticsearch is doing it when it's indexing
a document.

This also introduces a minor limitation: aliases type of fields that
are NOT part of a tree of sub-fields will not be able to be retrieved
anymore. field_caps API doesn't shed any light into a field being an
alias or not and at _source parsing time there is no way to know if a
root field is an alias or not. Fields of the type "a.b.c.alias" can be
extracted from docvalue_fields, only if the field they point to can be
extracted from docvalue_fields. Also, not all fields in a hierarchy of
fields can be evaluated to being an alias.

elasticmachine commented:
Pinging @elastic/es-search

astefan committed:
Fix tests

astefan committed:
More tests updates

astefan commented:
@elasticmachine run elasticsearch-ci/1

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java:
```diff
@@ -165,12 +173,46 @@ private Object unwrapMultiValue(Object values) {
+                    switch (dataType) {
+                        case BYTE:
+                            if (l > Byte.MAX_VALUE || l < Byte.MIN_VALUE) {
+                                return l;
```
Could you please double check if these ifs are needed (in terms of the error message produced, since the parse() itself does internally this validation)?
If they're still valid, please add a comment to indicate this.

matriv commented on docs/reference/sql/limitations.asciidoc:
```diff
@@ -161,3 +161,11 @@ By default,`geo_points` fields are indexed and have doc values. However only lat
+[[fields-from-source]]
+=== Retrieving from _source
+
+Most of {es-sql}'s columns are retrieved from `_source`, even though the mapping may disable the <<mapping-source-field,`_source`>> field.
```
This phrasing seems a bit confusing to me. I'd change this towards something like:
```
Most of {es-sql}'s columns are retrieved from `_source`, except from the fields of types: ..... 
``` (maybe list them with their SQL name instead of the ES name)
```If _source is disabled in the mapping and a column that must be retrieved from _source is requested in the SELECT .....```

matriv reviewed (commented):
Nice work overall!
Left some comments but nothing major.

Also, is there a need to have the `isAggregatable` as an attribute of `EsField` class? I was thinking it can just delegate to dataType.hasDocValues(), or there is some distinction between them?

astefan commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:
```diff
@@ -0,0 +1,424 @@
+            expectBadRequest(() -> {
+                client().performRequest(buildRequest(query));
+                return Collections.emptyMap();
+            }, containsString("unable to fetch fields from _source field: _source is disabled in the mappings for index [test]"));
```
This is actually an error message from Elasticsearch itself, it's not something SQL responds back with.

astefan commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java:
```diff
@@ -0,0 +1,424 @@
+        }
+    }
+    
+    private JDBCType jdbcTypeFor(String esType) {
```
Unfortunately, no.  `DataType` and `DataTypes` (belonging to `sql` project) are not visible in the `qa` project because `sql` is not a dependency. And `TypeUtils` is package protected.

jtibshirani commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java:
```diff
@@ -520,9 +522,26 @@ public void resolveAsSeparateMappings(String indexWildcard, String javaRegex, bo
                         if (field == null || (invalidField != null && (field instanceof InvalidMappedField) == false)) {
+                            int dot = fieldName.lastIndexOf('.');
+                            /*
+                             * Looking up the "tree" at the parent fields here to see if the field is an alias.
```
We discussed offline that this isn't a very solid way of checking whether a field is an alias. The fact that we don't return parent information about field aliases is kind of an 'accident' of the field caps logic, and I don't think it was an intentional API decision.

I filed #44298 to propose adding explicit field alias information to the field caps API, which would help make this more robust. It might also give enough information to be able to resolve + extract the alias field values from `_source`.

astefan committed:
Addressing review and adding more tests

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 41852_fix

astefan committed:
Limitations page update

astefan commented:
@matriv thank you for the review. I've addressed those and added more tests and covered some more scenarios in code. Would you mind having a second look, please?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java:
```diff
@@ -297,16 +298,36 @@ private String aliasName(Attribute attr) {
+        // Only if the field is not an alias (in which case it will be taken out from docvalue_fields if it's isAggregatable()),
+        // go up the tree of parents until a non-object (and non-nested) type of field is found and use that specific parent
+        // as the field to extract data from, from _source. We do it like this because sub-fields are not in the _source, only
+        // the root field to which those sub-fields belong to, are.
```
Would be great to add also an example with a subfield so it's more clear for someone that reads it in the future.

matriv reviewed (commented):
Awesome work with the extensive testing and code comments.
Left a few more comments.

matriv commented:
@astefan Should we also add a sentence in limitations page to say that if there are many docvalue fields (like date) a use can still experience the issue when doing a `SELECT *` ?

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java:
```diff
@@ -84,6 +88,7 @@ public FieldHitExtractor(String name, DataType dataType, ZoneId zoneId, boolean
 
     FieldHitExtractor(StreamInput in) throws IOException {
         fieldName = in.readString();
+        fullFieldName = in.readOptionalString();
```
I don't think we have this in place in other place in code? I mean, at the moment we enforce strict version matching for jdbc and, I think, for cli as well....

matriv replied:
But this is not for server-client mismatch but for version mismatch between nodes during a rolling upgrade.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java:
```diff
@@ -182,31 +200,21 @@ private Object unwrapMultiValue(Object values) {
-                            return l;
+                    Number result = null;
+                    try {
+                        result = NumberType.valueOf(dataType.esType.toUpperCase(Locale.ROOT)).parse(values, true);
```
Which part are you referring to?

matriv replied:
`NumberType.valueOf(dataType.esType.toUpperCase(Locale.ROOT))`
so it's a class method of a DataType and one can do:
`dataType.numberType()`

matriv replied:
But, please, it's minor and future oriented in case we want to reuse it somewhere else.

astefan committed:
Adressing reviews

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 41852_fix

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/container/QueryContainer.java:
```diff
@@ -300,12 +300,23 @@ private String aliasName(Attribute attr) {
-        // the root field to which those sub-fields belong to, are.
+        // the root field to which those sub-fields belong to, are. Instead of "text_field.keyword_subfield" for _source extraction,
+        // we use "text_field", because there is no source for "keyword_subfield".
+        /*
```
great, thx!

matriv reviewed (approved):
LGTM. Once again: great work!

astefan committed:
Add backwards compatibility checks for newly added serialized info

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 41852_fix

astefan commented:
@elasticmachine run elasticsearch-ci/2

astefan committed:
Small updates

astefan committed:
Merge branch 'master' of https://github.com/elastic/elasticsearch into 41852_fix

astefan committed:
Fix test
